Update PromQL from upstream 00ee73e: 28th Aug,20

Harkishen-Singh opened the pull request:
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

This PR updates PromQL and Promb from upstream as of commit 00ee73e: 28th Aug,20


Note: I am facing issues with `go mod vendor` since it was saying `github.com/prometheus/prometheus v2.5 +incompatible`. So, there remains 12 syntax errors in `engine.go` which is part of updated prometheus code, but due to incompatibility, it could not be copied to vendor. Any help related to this is appreciated.

Update: The above issue of incompatibility is now fixed!

cevian commented on pkg/prompb/remote.pb.go:
```diff
@@ -68,7 +68,7 @@ type WriteRequest struct {
 	XXX_sizecache        int32        `json:"-"`
 }
 
-func (m *WriteRequest) Reset()         { *m = WriteRequest{Timeseries: m.Timeseries[:0]} }
```
These are our changes. They shouldn't be backed out. Wonder if we should move these function to a different go file.

Harkishen-Singh replied:
Yeah, moving to a different file will be safer. I will keep our custom changes in `custom.pb.go` in the same dir so that it remains safe to know when changes happen.

cevian commented on pkg/prompb/remote.pb.go:
```diff
@@ -937,12 +937,7 @@ func (m *WriteRequest) Unmarshal(dAtA []byte) error {
-			} else {
-				m.Timeseries = append(m.Timeseries, TimeSeries{})
-			}
+			m.Timeseries = append(m.Timeseries, TimeSeries{})
```
Our changes should not be changed back

cevian commented on pkg/prompb/types.pb.go:
```diff
@@ -146,7 +146,7 @@ type TimeSeries struct {
 }
 
-func (m *TimeSeries) Reset()         { *m = TimeSeries{Labels: m.Labels[:0], Samples: m.Samples[:0]} }
+func (m *TimeSeries) Reset()         { *m = TimeSeries{} }
```
our intentional change. should not be undone.

cevian commented on pkg/prompb/types.pb.go:
```diff
@@ -255,7 +255,7 @@ type Labels struct {
 }
 
-func (m *Labels) Reset()         { *m = Labels{Labels: m.Labels[:0]} }
+func (m *Labels) Reset()         { *m = Labels{} }
```
our intentional change. should not be undone.

cevian commented on pkg/prompb/types.pb.go:
```diff
@@ -1384,12 +1384,7 @@ func (m *TimeSeries) Unmarshal(dAtA []byte) error {
-			} else {
-				m.Labels = append(m.Labels, Label{})
-			}
+			m.Labels = append(m.Labels, Label{})
```
our intentional change. should not be undone.

cevian commented on pkg/prompb/types.pb.go:
```diff
@@ -1423,12 +1418,7 @@ func (m *TimeSeries) Unmarshal(dAtA []byte) error {
-			} else {
-				m.Samples = append(m.Samples, Sample{})
-			}
+			m.Samples = append(m.Samples, Sample{})
```
our intentional change. should not be undone.

cevian commented on pkg/prompb/types.pb.go:
```diff
@@ -1634,12 +1624,7 @@ func (m *Labels) Unmarshal(dAtA []byte) error {
-			} else {
-				m.Labels = append(m.Labels, Label{})
-			}
+			m.Labels = append(m.Labels, Label{})
```
our intentional change. should not be undone.

cevian reviewed (changes requested):
Some changes override out intentional changes and should be undone. But mostly this looks ok

Harkishen-Singh commented:
Ahh, I just copied from upstream for promb assuming that it was generated so we wouldn't have edited it. I just kept in mind to make sure our intentional changes in promql should not be reverted.

Anyways, will revert the places where you have commented in the review.

Harkishen-Singh commented on pkg/promql/test.go:
```diff
@@ -78,8 +78,8 @@ type QuerierWrapper struct {
 func (t *QuerierWrapper) Select(b bool, sh *storage.SelectHints, _ []parser.Node, m ...*labels.Matcher) (storage.SeriesSet, parser.Node, storage.Warnings, error) {
-	ss, w, err := t.Querier.Select(b, sh, m...)
-	return ss, nil, w, err
+	ss := t.Querier.Select(b, sh, m...)
+	return ss, nil, nil, nil
```
@cevian pls see if this is okay.

Harkishen-Singh commented:
The `storage.SeriesSet` design has been changed in upstream. Now, it returns only SeriesSet (unexpandedseriesset). I have not made those changes in our code, rather worked around it to skip that change as it will be breaking on our side. There are some points (our way of `q.Select` in `series.go` I guess) to be discussed before doing those changes.

Harkishen-Singh commented:
Fixed all tests! Up for depth review especially for the querier.Select. Added promql testsdata from upstream that was missing.

Any idea why `go generate` is failing?

cevian commented on pkg/promql/test.go:
```diff
@@ -568,10 +568,11 @@ func (t *Test) clear() {
+		Reg:                      nil,
+		MaxSamples:               10000,
+		Timeout:                  100 * time.Second,
+		NoStepSubqueryIntervalFn: func(int64) int64 { return durationMilliseconds(1 * time.Minute) },
```
We should maybe add an feature-request issue to make this configurable

Harkishen-Singh replied:
Ahh, can you explain this bit more? Is this for upstream?

cevian commented on pkg/pgmodel/sql_reader.go:
```diff
@@ -86,15 +86,14 @@ func (q *pgxQuerier) LabelsCacheCapacity() int {
-
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, nil
```
It seems like this error is getting lost. Should we return an errorSeriesSet?

Harkishen-Singh replied:
yeah, that will be better. I wonder how I missed that.

cevian replied:
This still needs to be done

cevian commented on pkg/api/series.go:
```diff
@@ -71,12 +71,12 @@ func series(queryable *query.Queryable) http.HandlerFunc {
-				return
-			}
+			s, _ := q.Select(false, nil, nil, mset...)
+			//warnings = append(warnings, wrn...)
```
can't we grab these warning from s.Warnings() now?

Harkishen-Singh replied:
Ahh, I missed those debug comments. Anyways, the warnings will now be handled in the SeriesSet. I kept Warnings as nil in pgxSeriesSet since in upstream, the `blockBaseSeriesSet` has the same behaviour.

But still, being on the safe side, we can make a way to pass warnings through `(*pgxSeriesSet).Warnings()`.

cevian replied:
Yes let's do that. (pass it up)

JLockerman replied:
(even if we do only ever return `nil` in our current `SeriesSet` we should still retain the logic to forward the warnings here for ease of future changes)

Harkishen-Singh replied:
Agreed.

cevian commented on pkg/prompb/custom.pb.go:
```diff
@@ -0,0 +1,5 @@
+package prompb
```
Like these a lot.

cevian commented on pkg/prompb/remote.pb.go:
```diff
@@ -68,7 +68,6 @@ type WriteRequest struct {
 	XXX_sizecache        int32        `json:"-"`
 }
 
```
I'd maybe leave a comment here that the Reset function was moved to custom.pb.go. Same for all places it moved

cevian commented on pkg/prompb/remote.proto:
```diff
@@ -0,0 +1,82 @@
+// Copyright 2016 Prometheus Team
```
I'm not sure we should move the proto files in here. It is incorrect to try to recompile them into Go because of our optimizations.

Harkishen-Singh replied:
In that case, should I let them be as they were?

cevian commented on pkg/prompb/rpc.proto:
```diff
@@ -0,0 +1,76 @@
+// Copyright 2017 Prometheus Team
```
same as above for .proto files

cevian commented on pkg/prompb/types.proto:
```diff
@@ -0,0 +1,85 @@
+// Copyright 2017 Prometheus Team
```
this file too for .proto

cevian commented on pkg/promql/engine.go:
```diff
@@ -743,22 +728,17 @@ func (ng *Engine) populateSeries(ctx context.Context, querier Querier, s *parser
-				return err
+				return fmt.Errorf("assigning topNode twice")
 			}
+			set, topNode = querier.Select(false, hints, path, n.LabelMatchers...)
```
hmm (for another PR) should we consider making topNode also part of the SeriesSet?

Harkishen-Singh replied:
We anyhow need to keep topNode in evaluator for checks in `ev.eval()`. (I think so, not sure).

Harkishen-Singh replied:
Sure, will open an issue for this.

Harkishen-Singh commented:
@cevian made changes as per the planings on this PR.

cevian commented on pkg/api/series.go:
```diff
@@ -71,12 +71,12 @@ func series(queryable *query.Queryable) http.HandlerFunc {
-			}
+			s, _ := q.Select(false, nil, nil, mset...)
+			//warnings = append(warnings, wrn...)
+			//if err != nil {
```
We should probably check s.Err() here? and not just comment out?

Harkishen-Singh replied:
This is addressed. Hence, resolving.

cevian reviewed (changes requested):
A few small things. will ask for a review from Josh + Ante

JLockerman commented on pkg/api/marshal.go:
```diff
@@ -43,7 +43,7 @@ func marshalCommonFooter(out *errorWrapper, warnings []string) {
 
 func marshalMatrixData(out *errorWrapper, data promql.Matrix) {
-	out.WriteStrings(`{"resultType":"`, parser.ValueTypeMatrix, `","result":[`)
+	out.WriteStrings(`{"resultType":"`, string(parser.ValueTypeMatrix), `","result":[`)
```
Please don't change code independant of the PR at hand: it makes it more difficult to review PRs, and to track down where changes happen.

Harkishen-Singh replied:
This PR has two commits: 1st for changes related to just PromQL and the 2nd for changes that are dependent on the PromQL update. This line is part of the 2nd commit.

JLockerman commented on go.mod:
```diff
@@ -3,30 +3,78 @@ module github.com/timescale/timescale-prometheus
+	github.com/hashicorp/golang-lru v0.5.4 // indirect
+	github.com/hashicorp/serf v0.9.4 // indirect
+	github.com/imdario/mergo v0.3.5 // indirect
+	github.com/influxdata/influxdb v1.8.2 // indirect
```
I don't love this `go.mod change` is there a way we can shrink it?

JLockerman replied:
There doesn't seem to be any significant new dependencies...

Harkishen-Singh replied:
I am not sure that manual intervention will do this. Any suggestions?

JLockerman commented on pkg/pgmodel/series_set.go:
```diff
@@ -99,6 +99,9 @@ func (p *pgxSeriesSet) Err() error {
 	return nil
 }
 
+// Warnings implements storage.SeriesSet.
```
this comment is likely wrong

Harkishen-Singh replied:
Wrong in sense?

JLockerman reviewed (commented):
Besides what's mentioned in earlier comments, the changes look good!
The only other big thing is that these need real commit messages: at a minimum the first commit needs to describe _why_ we're doing this upgrade, and what the commit range we're including is; the second should describe the kinds of changes necessitated by the new code we're merging in ([this article](https://chris.beams.io/posts/git-commit/) may be helpful, seems to be what the DB team is using now).

antekresic commented on pkg/api/series.go:
```diff
@@ -71,9 +71,9 @@ func series(queryable *query.Queryable) http.HandlerFunc {
+			s, _ := q.Select(false, nil, nil, mset...)
+			warnings = append(warnings, s.Warnings()...)
+			if s.Err() != nil {
 				respondError(w, http.StatusUnprocessableEntity, err, "execution")
```
Should this be changed to `s.Err()` instead of `err`?

Harkishen-Singh replied:
Missed that! Addressed now.

antekresic reviewed (dismissed):
Seems like the tests are failing, I believe its connected to my comment. After that is fixed, should be good to go, along with the commit message changes that @JLockerman mentioned previously.
